src/main/scala/org/broadinstitute/dsde/firecloud/service/LibraryService.scala

```diff
@@ -54,14 +57,52 @@ class LibraryService (protected val argUserInfo: UserInfo, val rawlsDAO: RawlsDA
   implicit val userInfo = argUserInfo
 
   override def receive = {
-    case UpdateAttributes(ns: String, name: String, attrsJsonString: String) => asCurator {updateAttributes(ns, name, attrsJsonString)} pipeTo sender
-    case SetPublishAttribute(ns: String, name: String, value: Boolean) => asCurator {setWorkspaceIsPublished(ns, name, value)} pipeTo sender
+    case UpdateAttributes(ns: String, name: String, attrsJsonString: String) => updateAttributes(ns, name, attrsJsonString) pipeTo sender
+    case UpdateDiscoverableBy(ns: String, name: String, newGroups: Seq[String]) => updateDiscoverableBy(ns, name, newGroups) pipeTo sender
+    case SetPublishAttribute(ns: String, name: String, value: Boolean) => setWorkspaceIsPublished(ns, name, value) pipeTo sender
     case IndexAll => asAdmin {indexAll} pipeTo sender
     case FindDocuments(criteria: LibrarySearchParams) => findDocuments(criteria) pipeTo sender
     case Suggest(criteria: LibrarySearchParams) => suggest(criteria) pipeTo sender
-    case PopulateSuggest(field: String, text: String) => asCurator {populateSuggest(field: String, text: String)} pipeTo sender
+    case PopulateSuggest(field: String, text: String) => populateSuggest(field: String, text: String) pipeTo sender
   }
 
+  def hasAccessOrCurator(workspaceResponse: WorkspaceResponse, neededLevel: WorkspaceAccessLevels.WorkspaceAccessLevel): Future[Boolean] = {
+    val wsaccess = workspaceResponse.accessLevel >= neededLevel
+    if (!wsaccess)
+      tryIsCurator(userInfo)
+    else
+      Future.successful(wsaccess)
+  }
+
+  def isPublished(workspaceResponse: WorkspaceResponse): Boolean = {
+    workspaceResponse.workspace.attributes.get(publishedFlag).fold(false)(_.asInstanceOf[AttributeBoolean].value)
+  }
+
+  /*
+   * This attribute is allowed to be modified by on owner or someone with canShare for the workspace
+   */
+  def updateDiscoverableBy(ns: String, name: String, newGroups: Seq[String]): Future[PerRequestMessage] = {
+    if (FireCloudConfig.ElasticSearch.discoverGroupNames.containsAll(newGroups.asJavaCollection)) {
+      rawlsDAO.getWorkspace(ns, name) map { workspaceResponse =>
+        if (workspaceResponse.accessLevel >= WorkspaceAccessLevels.Owner) {
+          // this is technically vulnerable to a race condition in which the workspace attributes have changed
+          // between the time we retrieved them and here, where we update them.
+          val remove = Seq(RemoveAttribute(discoverableWSAttribute))
+          val operations = newGroups map (group => AddListMember(discoverableWSAttribute, new AttributeString(group)))
+          RequestComplete(patchWorkspace(ns, name, remove ++ operations, isPublished(workspaceResponse)))
+        } else {
+          RequestCompleteWithErrorReport(Forbidden, "must be an owner to set discoverable By")
```

Comment:
nitpick: `Must be an owner to set the groups this workspace is discoverable by.` seems better. My inner grammar police is screaming not to end a sentence with a preposition, but this is considered OK now.